feat/AF-186: Add keytar for secure token storage with file-based fallback

LucasR29 opened the pull request:
This commit introduces keytar as a primary method for securely storing
API tokens using the system's native credential storage. It maintains
the existing file-based storage as a fallback option for compatibility
and robustness across different environments.

This commit introduces keytar as a primary method for securely storing API tokens using the system's native credential storage. It also creates a  file-based storage as a fallback option for compatibility and robustness across different environments.

- Implement keytar for storing and retrieving tokens
- Implement storeToken and getToken functions to safely store the api key
- Fallback to file-based storage if keytar fails
- Update error handling and logging for both storage methods
- Fix ts configs

**TESTING**

```
pnpm install
```
```
pnpm build
```
```
pnpm anyflow init
```
```
pnpm anyflow auth
```

yudikubota commented on src/cli.ts:
```diff
@@ -1,33 +1,78 @@
 
-console.log(figlet.textSync("Anyflow CLI"));
+async function ensureEnvFile() {
+  const envPath = path.resolve(process.cwd(), '.env');
```
`process.cwd()` will return the current directory the cli is running on, not the project's current directory. Example of how this can be an issue:

```
cd my-project/subdirectory
anyflow init
```

the `.env` file will be stored on `my-project/subdirectory/.env` and not on the project root

yudikubota replied:
Okay solved 👍🏻

yudikubota commented on src/cli.ts:
```diff
@@ -1,33 +1,78 @@
+    await fs.access(envPath);
+  } catch {
+    const key = crypto.randomBytes(32).toString('hex');
+    await fs.writeFile(envPath, `ENCRYPTION_KEY=${key}`);
```
`writeFile` will overwrite all `.env` content if there's any already there. I think you should only append the content.
Also, it is possible you have access to the file, but the ENCRYPTION_KEY variable is not defined.

yudikubota commented on src/cli.ts:
```diff
@@ -1,33 +1,78 @@
+}
 
-const program = new Command();
+async function init() {
```
Is `init()` command necessary? I think it's not.

By default, `main()` function is called and the first thing it does is call the same functions as `init()` does.

yudikubota commented on src/commands/auth.ts:
```diff
@@ -1,32 +1,120 @@
+    console.log("Browser opened successfully.");
+  } catch (err) {
+    console.error(`Error opening URL: ${err}`);
+    process.exit(1);
```
If the `open()` command fails the user will always fall into catch handler and exit the process, making it impossible to log in.

I prefer doing something like other CLIs: Show the user what URL we're trying to open and give the option for them to open themselves if the open() command fails.

yudikubota commented on src/commands/auth.ts:
```diff
@@ -1,32 +1,120 @@
+  } finally {
     rl.close();
-  });
+  }
```
I think right after storing the token we should make an API call to test if it works. You can call `GET /api/user` to get user info and store into keytar's ACCOUNT.

LucasR29 replied:
Added

yudikubota commented on src/commands/auth.ts:
```diff
@@ -1,32 +1,120 @@
+  }
+}
+
+export async function checkAuth() {
```
similarly, checkAuth() should make a request to a backend endpoint to make sure the token works.

LucasR29 replied:
Added too

yudikubota replied:
This command has failed when the token is stored in file

![image](https://github.com/user-attachments/assets/d53772f9-13d4-4477-8f68-4556321bd919)

yudikubota commented on package.json:
```diff
@@ -26,6 +29,7 @@
     "@types/figlet": "^1.7.0",
     "@types/node": "^22.5.3",
     "@types/openurl": "^1.0.3",
+    "electron": "^32.2.0",
```
is electron being used?

LucasR29 replied:
I used to do a test, removed

yudikubota commented on src/commands/auth.ts:
```diff
@@ -1,32 +1,120 @@
 const ACCOUNT_NAME = "apiToken";
+const ENCRYPTION_KEY = process.env.ENCRYPTION_KEY;
+
+if (!ENCRYPTION_KEY) {
```
This is running at the moment you import auth.js and is preventing us of using the tool. This check is unnecessary as to get to the functions here you would need to pass through `main()` function which already checks the encryption key.

Steps to reproduce:
```
anyflow init
rm .env
anyflow auth
anyflow init
```

we get:

![image](https://github.com/user-attachments/assets/18b4907f-476d-412b-8b3f-391008476b6b)

impossible to init()

LucasR29 replied:
Fixed, the problem was on the auth.ts file that made the verification outside the function

LucasR29 commented:
- Improved folder structure
- Added validation on route ```/api/user```
- Created tutorial file for keytar configuration, it is neded to configure it on the project folder, at least on my machine.
-  Added comments
- Tried to add User to the keytar account, but that is not possible without saving the user, if allowed I will can save it on the .env file
- Changed ```cwd``` to ```pkg-dir```, tested using ```npm link``` 
- Changed ```function main``` so it dosent execute the init functions
- Changed ```write file``` to ```append file```

yudikubota commented on src/commands/auth/store-token/store.ts:
```diff
@@ -0,0 +1,76 @@
+
+    console.log("Token and user info stored securely using system keychain.");
+  } catch (error) {
+    console.error("Error storing token:", error);
```
This prints out the whole error when we input an invalid token. This is not good UX. You should handle the invalid token errors and provide a short feedback to the user.

![image](https://github.com/user-attachments/assets/a66fd9aa-5968-4b45-bc44-3441a4b0ebca)

yudikubota commented on src/commands/auth/api/user.ts:
```diff
@@ -0,0 +1,34 @@
+
+// Make a request to get user information
+export async function getUserResponse(token:string) {
+    const response = await axios.get(`${BACKEND_URL}/user`, {
```
if axios fails it will throw. 
but status handling expects it to not throw.

please test all possible statuses using https://httpstat.us/

yudikubota reviewed (changes requested):
check-auth is not working when api key is stored in file
also need to improve UX and error handling for axios http errors

yudikubota reviewed (changes requested):
We're still getting unhandled errors in axios when we input an invalid key. The system should provide the user with feedback on what went wrong, what are the causes and how to proceed. This is an example of a good feedback message:

"There was an error when testing your API Key: 401 - Unauthenticated. Please make sure it is a valid key and try again."

![image](https://github.com/user-attachments/assets/0474ce84-43d8-4dc0-b089-d790d680896a)

To be clear. The CLI user don't need to know we're using axios and do not want to debug it himself. Let's focus on UX as it's one of the main focuses of AnyFlow.

yudikubota reviewed (approved):
Actually I forgot to build the project. Everything is fine. Approved!
